Poll generic process parking at x86 syscall entry

--- a/arch/x86/entry/syscall_64.c
+++ b/arch/x86/entry/syscall_64.c
@@ -9,6 +9,7 @@
 #include <linux/nospec.h>
 #include <asm/syscall.h>
 #include <asm/harmony_pvclock.h>
+#include <linux/harmony_park.h>
 
 #define __SYSCALL(nr, sym) extern long __x64_##sym(const struct pt_regs *);
 #define __SYSCALL_NORETURN(nr, sym) extern long __noreturn __x64_##sym(const struct pt_regs *);
@@ -98,6 +99,8 @@
 
 	instrumentation_begin();
 
+	harmony_park_poll();
+
 	if (!do_syscall_x64(regs, nr) && !do_syscall_x32(regs, nr) && nr != -1) {
 		/* Invalid system call, but still a system call. */
 		regs->ax = __x64_sys_ni_syscall(regs);
